src/libraries/System.Text.Json/tests/Serialization/EnumConverterWithNamingPolicyTests.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Text.Json.Serialization.Tests
+{
+   public class EnumConverterWithNamingPolicyTests
+    {
+        private readonly ITestOutputHelper _outputHelper;
+
+        public EnumConverterWithNamingPolicyTests(ITestOutputHelper outputHelper)
+        {
+            _outputHelper = outputHelper;
+        }
+
+        public class SnakeCaseNamingPolicy : JsonNamingPolicy
+        {
+            public override string ConvertName(string name)
+            {
+                if (name == null)
+                {
+                    throw new ArgumentNullException(nameof(name));
+                }
+                var result = new StringBuilder();
+                for (var i = 0; i < name.Length; i++)
+                {
+                    var c = name[i];
+                    if (i == 0)
+                    {
+                        result.Append(char.ToLower(c));
+                    }
+                    else
+                    {
+                        if (char.IsUpper(c))
+                        {
+                            result.Append('_');
+                            result.Append(char.ToLower(c));
+                        }
+                        else
+                        {
+                            result.Append(c);
+                        }
+                    }
+                }
+                return result.ToString();
+            }
+
+        }
+        public enum TestType
+        {
+            None,
+            ValueOne,
+            ValueTwo,
+        }
+
+        public class ObjectWithEnumProperty
+        {
+            public TestType TestType { get; set; }
+        }
+
+        [Fact]
+        public void TestEnumCase()
+        {
+            var namingPolicy = new SnakeCaseNamingPolicy();
+
+            var opts = new JsonSerializerOptions()
+            {
+                PropertyNamingPolicy = namingPolicy,
+                DictionaryKeyPolicy = namingPolicy,
+                Converters =
+                {
+                    new JsonStringEnumConverter(namingPolicy)
+                }
+            };
+
+            var enumValues = Enum.GetValues(typeof(TestType)).Cast<TestType>();
+            foreach (var v in enumValues)
+            {
+                var sourceObject = new ObjectWithEnumProperty()
+                {
+                    TestType = v
+                };
+
+                var json = JsonSerializer.Serialize(sourceObject, opts);
+                _outputHelper.WriteLine(json);
+                var deserializedObject = JsonSerializer.Deserialize<ObjectWithEnumProperty>(json, opts);
```

Comment:
If `json` were not serialized by `JsonSerializer` with `opts` would this still work?

```csharp
[Theory]
[InlineData(@"""none""", TestType.None)]
[InlineData(@"0", TestType.None)]
[InlineData(@"""value_one""", TestType.ValueOne)]
[InlineData(@"1", TestType.ValueOne)]
[InlineData(@"""value_two""", TestType.ValueTwo)]
[InlineData(@"2", TestType.ValueTwo)]
public void TestEnumNamingPolicy(string json, TestType expectedValue)
{
    // ...opts as above
    string json = $@"{{ ""test_type"": {json} }}";
    var deserializedObject = JsonSerializer.Deserialize<ObjectWithEnumProperty>(json, opts);
    Assert.Equal(expectedValue, deserializedObject.TestType);
}
```

I ask because it appears that the cache for enum names is only ever populated if you write the value first.

Reply:
Cache is initialized by EnumConverter constructor 
https://github.com/dotnet/runtime/pull/42302/files/2a7699d12692176a00de0d3892960d383a13e2c5#diff-1836df549fbe8f438e0cfc38f3fc07daR76-R77 with only condition that naming policy exists.
Test cases will work if opts contains a JsonStringEnumConverter with snake case naming policy
https://github.com/dotnet/runtime/pull/42302/files/2a7699d12692176a00de0d3892960d383a13e2c5#diff-212a88a736cc2af2d4071f394407acd5R71-R79

Reply:
Thanks, I was unable to get it work pass locally, looks like it may be an issue on my end!

Reply:
Sory, i was wrong, cache initializer will be for first 64 enum values.
https://github.com/dotnet/runtime/pull/42302/files/2a7699d12692176a00de0d3892960d383a13e2c5#diff-1836df549fbe8f438e0cfc38f3fc07daR59-R64